### PR TITLE
Add locking around ops that change template router state.

### LINKED ISSUE #8825: router can hit concurrent map read and map write during commit
```
I0510 19:28:30.398672       1 plugin.go:173] Modifying routes for alt/altService
fatal error: concurrent map read and map write

goroutine 88 [running]:
runtime.throw(0x3a8f6a0, 0x21)
	/usr/local/go/src/runtime/panic.go:530 +0x90 fp=0xc8204f5218 sp=0xc8204f5200
runtime.mapaccess2(0x29197c0, 0xc82060dd10, 0xc8200112e0, 0xc8204f52c8, 0xc8206a3760)
	/usr/local/go/src/runtime/hashmap.go:343 +0x5a fp=0xc8204f5260 sp=0xc8204f5218
reflect.mapaccess(0x29197c0, 0xc82060dd10, 0xc8200112e0, 0xc82060dd10)
	/usr/local/go/src/runtime/hashmap.go:993 +0x35 fp=0xc8204f5290 sp=0xc8204f5260
reflect.Value.MapIndex(0x29197c0, 0xc8204d4320, 0x95, 0x29219c0, 0xc8200112e0, 0x98, 0x0, 0x0, 0x0)
	/usr/local/go/src/reflect/value.go:1041 +0x14a fp=0xc8204f5318 sp=0xc8204f5290
text/template.(*state).walkRange(0xc8204f58a0, 0x3402dc0, 0xc8204d4310, 0x99, 0xc820607000)
	/usr/local/go/src/text/template/exec.go:327 +0x885 fp=0xc8204f5518 sp=0xc8204f5318
text/template.(*state).walk(0xc8204f58a0, 0x3402dc0, 0xc8204d4310, 0x99, 0x7f4bd442f070, 0xc820607000)
	/usr/local/go/src/text/template/exec.go:226 +0x264 fp=0xc8204f5660 sp=0xc8204f5518
text/template.(*state).walk(0xc8204f58a0, 0x3402dc0, 0xc8204d4310, 0x99, 0x7f4bd443a190, 0xc820608630)
	/usr/local/go/src/text/template/exec.go:223 +0x6a1 fp=0xc8204f57a8 sp=0xc8204f5660
text/template.(*Template).Execute(0xc820610a80, 0x7f4bd43f3370, 0xc8200667b8, 0x3402dc0, 0xc8204d4310, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:178 +0x2a7 fp=0xc8204f58e8 sp=0xc8204f57a8
github.com/openshift/origin/pkg/router/template.(*templateRouter).writeConfig(0xc8204c48f0, 0x0, 0x0)
	/go/src/github.com/openshift/origin/pkg/router/template/router.go:284 +0x892 fp=0xc8204f5d48 sp=0xc8204f58e8
github.com/openshift/origin/pkg/router/template.(*templateRouter).commitAndReload(0xc8204c48f0, 0x0, 0x0)
	/go/src/github.com/openshift/origin/pkg/router/template/router.go:232 +0x14d fp=0xc8204f5db0 sp=0xc8204f5d48
github.com/openshift/origin/pkg/router/template.(*templateRouter).(github.com/openshift/origin/pkg/router/template.commitAndReload)-fm(0x0, 0x0)
	/go/src/github.com/openshift/origin/pkg/router/template/router.go:158 +0x2e fp=0xc8204f5dd0 sp=0xc8204f5db0
github.com/openshift/origin/pkg/util/ratelimiter.(*RateLimitedFunction).handleOne(0xc82060dc20, 0x3491820, 0xc82060dc20)
	/go/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter.go:49 +0x48 fp=0xc8204f5e08 sp=0xc8204f5dd0
github.com/openshift/origin/pkg/util/ratelimiter.(*RateLimitedFunction).RunUntil.func1()
	/go/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter.go:42 +0x6a fp=0xc8204f5e50 sp=0xc8204f5e08
k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1(0xc82060e570)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wait/wait.go:66 +0x4b fp=0xc8204f5e80 sp=0xc8204f5e50
k8s.io/kubernetes/pkg/util/wait.JitterUntil(0xc82060e570, 0x0, 0x0, 0xc8200f0900)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wait/wait.go:67 +0x68 fp=0xc8204f5f40 sp=0xc8204f5e80
k8s.io/kubernetes/pkg/util/wait.Until(0xc82060e570, 0x0, 0xc8200f0900)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wait/wait.go:47 +0x3e fp=0xc8204f5f68 sp=0xc8204f5f40
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc8204f5f70 sp=0xc8204f5f68
created by github.com/openshift/origin/pkg/util/ratelimiter.(*RateLimitedFunction).RunUntil
	/go/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter.go:42 +0x7d

goroutine 1 [select (no cases)]:
github.com/openshift/origin/pkg/cmd/infra/router.(*TemplateRouterOptions).Run(0xc8203c8f00, 0x0, 0x0)
	/go/src/github.com/openshift/origin/pkg/cmd/infra/router/template.go:203 +0x85e
github.com/openshift/origin/pkg/cmd/infra/router.NewCommandTemplateRouter.func1(0xc8203e2600, 0xc8204c23a0, 0x0, 0x2)
	/go/src/github.com/openshift/origin/pkg/cmd/infra/router/template.go:111 +0x1ad
github.com/spf13/cobra.(*Command).execute(0xc8203e2600, 0xc8200600d0, 0x2, 0x2, 0x0, 0x0)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/github.com/spf13/cobra/command.go:572 +0x85a
github.com/spf13/cobra.(*Command).ExecuteC(0xc8203e2600, 0xc8203e2600, 0x0, 0x0)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/github.com/spf13/cobra/command.go:662 +0x53f
github.com/spf13/cobra.(*Command).Execute(0xc8203e2600, 0x0, 0x0)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/github.com/spf13/cobra/command.go:618 +0x2d
main.main()
	/go/src/github.com/openshift/origin/cmd/openshift/openshift.go:27 +0x180
```
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/808/consoleFull

### PR DESCRIPTION
**Pull request by ramr**
fixes #8825 

@liggitt   PTAL

### CODE REVIEW

File: pkg/router/template/router.go [hunk 1]
```diff
@@ -506,6 +527,8 @@ func (r *templateRouter) AddEndpoints(id string, endpoints []Endpoint) bool {
 		return false
 	}
 
+	r.lock.Lock()
```
Comment (smarterclayton) [inline pkg/router/template/router.go]: Find service unit does not make a copy, so this is not safe.

File: pkg/router/template/router.go [hunk 2]
```diff
@@ -488,6 +506,9 @@ func (r *templateRouter) RemoveRoute(id string, route *routeapi.Route) {
 	}
 
 	routeKey := r.routeKey(route)
+
+	r.lock.Lock()
```
Comment (smarterclayton) [inline pkg/router/template/router.go]: What about the read to r.state above?

File: pkg/router/template/router.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/router/template/router.go]: Unless we also guarantee that we make no mutations to state.  Is that the case?

File: pkg/router/template/router.go (hunk 2 above)
Comment (ramr) [inline pkg/router/template/router.go]: My bad - missed that.

File: pkg/router/template/router.go (hunk 1 above)
Comment (ramr) [inline pkg/router/template/router.go]: Hmm, might be better to separate FindServiceUnit into an internal/external part - that way all the internal calls can work thru' in a lock. Let me redo this. Thx

**Comment by ramr** [comment]
[test]

**Comment by smarterclayton** [comment]
Changes look good to me - will wait for clean run and then apply label.

**Comment by smarterclayton** [comment]
@deads2k you are pinged re: test basic controller workflow test

**Comment by deads2k** [comment]
re[test]

Another failure in the test flow, still haven't seen one in merge.  I think I'll try bumping the timeout.  Maybe someone's parallel test-cmd and test-go makes things extra slow.

**Comment by smarterclayton** [comment]
[merge] Lgtm

**Pull request merged**